text/2018-01-07-deque-proof.md

```diff
@@ -0,0 +1,994 @@
+    'L203: fence(SeqCst);
+    'L204: let t = self.top.load(Relaxed);
+    'L205: let len = b - t;
+
+    'L206: if len <= 0 {
+    'L207:     self.bottom.store(b, Relaxed);
+    'L208:     return None;
+    'L209: }
+
+    'L210: let buffer = self.buffer.load(Relaxed, epoch::unprotected());
+    'L211: let mut value = Some(buffer.read((b - 1) % buffer.get_capacity(), Relaxed));
+
+    'L212: if len == 1 {
+    'L213:     if self.top.compare_and_swap(t, t + 1, Release, Acquire).is_err() {
+    'L214:         mem::forget(value.take());
+    'L215:     }
+    'L216:     self.bottom.store(b, Relaxed);
+    'L217: } else {
+    'L218:     let cap = buffer.get_capacity();
+    'L219:     if len < cap / 4 {
+    'L220:         self.resize(cap / 2);
+    'L221:     }
+    'L221: }
+
+    'L222: value
+}
+
+fn resize(&self, cap_new) {
+    'L301: let b = self.bottom.load(Relaxed);
+    'L302: let t = self.top.load(Relaxed);
+    'L303: let old = self.buffer.load(Relaxed, epoch::unprotected());
+    'L304: let new = Buffer::new(cap_new);
+
+    'L305: ... // copy data from old to new
+
+    'L306: let guard = epoch::pin();
+    'L307: self.buffer.store(Owned::new(new).into_shared(guard), Release);
+    'L308: guard.defer(move || old.into_owned());
+}
+
+pub fn steal(&self) -> Option<T> {
+    'L401: let mut t = self.top.load(Relaxed);
+
+    'L402: let guard = epoch::pin_fence(); // epoch::pin(), but issue fence(SeqCst) even if it is re-entering
+    'L403: loop {
+    'L404:     let b = self.bottom.load(Acquire);
+    'L405:     if b - t <= 0 {
+    'L406:         return None;
+    'L407:     }
+
+    'L408:     let buffer = self.buffer.load(Acquire, &guard);
+    'L409:     let value = buffer.read(t % buffer.get_capacity(), Relaxed);
+
+    'L410:     match self.top.compare_and_swap_weak(t, t + 1, Release) {
+    'L411:         Ok(_) => return Some(value),
+    'L412:         Err(t_old) => {
+    'L413:             mem::forget(value);
+    'L414:             t = t_old;
+    'L415:             fence(SeqCst);
+    'L416:         }
+    'L417:     }
+    'L418: }
+}
+```
+
+
+## Safety
+
+We first prove that the proposed implementation is memory-safe, i.e. it does not invoke undefined
+behavior due to illegal memory accesses. We discuss two possible causes of undefined behavior: data
+race and invalid pointer dereference.
+
+
+### Absence of Data Races
+
+Every shared objects are atomically accessed so that there are no data races that invoke undefined
+behavior.
+
+In particular, the contents of the buffer is atomically accessed (`'L109`, `'L211`, `'L409`). This
+is necessary because `fn push(): 'L109` and `fn steal(): 'L409` may concurrently access the contents
+of the `buffer`, while the former is writing to it. For example, the scheduler may stop a `steal()`
+invocation right after `'L402` so that `t` read in `'L401` may be arbitrarily stale. Now, suppose
+that in a concurrent `push()` invocation, `b` equals to `t + buffer.get_capacity()` and it is
```

Comment:
Should this be `t + buffer.get_capacity() - 1`? Because if it is equal to `t + buffer.get_capacity()` that means the deque is full and the push operation cannot proceed.

Reply:
`push()` first writes to `buffer[b % sizeof(buffer)]` and then increases `bottom`, so the case I'd like to consider here is `b = t + sizeof(buffer)`.

It is possible in the presence of resizing, because the `top` variable can be advanced while the stealer is stalled by the scheduler, so that `t` (the old value) is much less than `top` (the current value).